Add encrypt-node status to cilium status

romanspb80 opened the pull request:
This patch adds param "NodeEncryption" to the cilium status

$ cilium status
[...]
Encryption:   Wireguard [NodeEncryption: true, cilium_wg0 (Pubkey: UzRSJUmRqsxl82xvFG0joBheaiqOnM3BCgpByYjVtCE=, Port: 51871, Peers: 3)]

Fixes: #24239

```release-note
cmd: Add NodeEncryption status to the cilium status command
```

nathanjsweet commented on cilium-health/cmd/get.go:
```diff
@@ -5,6 +5,7 @@ package cmd
 
 import (
 	"fmt"
+	"github.com/cilium/cilium/pkg/node"
```
Make sure to run `goimports` on this file so that it conforms to the linter.

nathanjsweet commented on pkg/client/client.go:
```diff
@@ -271,7 +271,7 @@ var (
 // whether a additional details are printed about a certain aspect of the status. In case there are
 // errors, some details may be printed regardless of the value of sd.
-func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetails) {
+func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetails, ne bool) {
```
Nit: Please make the variable name more descriptive.
For example:
```suggestion
func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetails, encryptStatus bool) {
```
Please also change invocations of this function to have the variable passed be something like "encryptStatus" as well.

nathanjsweet reviewed (changes requested):
Overall this looks good, but some minor fixes are needed. Thanks for the submission!

nathanjsweet commented on cilium-health/cmd/get.go:
```diff
@@ -31,12 +33,14 @@ var healthGetCmd = &cobra.Command{
 			fmt.Fprintf(w, "Node load:\t%s %s %s\n",
 				load.Last1min, load.Last5min, load.Last15min)
-			ciliumClient.FormatStatusResponse(w, &sr.Cilium, ciliumClient.StatusNoDetails)
+			ciliumClient.FormatStatusResponse(w, &sr.Cilium, ciliumClient.StatusNoDetails, statusNodeEncription)
```
```suggestion
			ciliumClient.FormatStatusResponse(w, &sr.Cilium, ciliumClient.StatusNoDetails, statusNodeEncryption)
```

nathanjsweet commented on cilium/cmd/debuginfo.go:
```diff
@@ -264,9 +265,11 @@ func addKernelVersion(w *tabwriter.Writer, p *models.DebugInfo) {
 	printMD(w, "Cilium status", "")
 	printTicks(w)
-	pkg.FormatStatusResponse(w, p.CiliumStatus, pkg.StatusAllDetails)
+	pkg.FormatStatusResponse(w, p.CiliumStatus, pkg.StatusAllDetails, statusNodeEncription)
```
```suggestion
	pkg.FormatStatusResponse(w, p.CiliumStatus, pkg.StatusAllDetails, statusNodeEncryption)
```

nathanjsweet commented on cilium/cmd/status.go:
```diff
@@ -89,7 +92,7 @@ func statusDaemon() {
 		sr := resp.Payload
 		w := tabwriter.NewWriter(os.Stdout, 2, 0, 3, ' ', 0)
-		pkg.FormatStatusResponse(w, sr, statusDetails)
+		pkg.FormatStatusResponse(w, sr, statusDetails, statusNodeEncription)
```
```suggestion
		pkg.FormatStatusResponse(w, sr, statusDetails, statusNodeEncryption)
```

nathanjsweet commented on pkg/client/client.go:
```diff
@@ -271,7 +271,7 @@ var (
 // whether a additional details are printed about a certain aspect of the status. In case there are
 // errors, some details may be printed regardless of the value of sd.
-func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetails) {
+func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetails, statusNodeEncription bool) {
```
```suggestion
func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetails, statusNodeEncryption bool) {
```

nathanjsweet commented on pkg/client/client.go:
```diff
@@ -689,8 +689,8 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetai
-				iface := fmt.Sprintf("%s (Pubkey: %s, Port: %d, Peers: %d)",
-					i.Name, i.PublicKey, i.ListenPort, i.PeerCount)
+				iface := fmt.Sprintf("NodeEncryption: %t, %s, (Pubkey: %s, Port: %d, Peers: %d)",
+					statusNodeEncription, i.Name, i.PublicKey, i.ListenPort, i.PeerCount)
```
```suggestion
					statusNodeEncryption, i.Name, i.PublicKey, i.ListenPort, i.PeerCount)
```

nathanjsweet reviewed (changes requested):
nit: typos

gandro commented on cilium-health/cmd/get.go:
```diff
@@ -31,12 +33,14 @@ var healthGetCmd = &cobra.Command{
 			}
 		} else {
+			//NodeEncription status
+			statusNodeEncryption := node.GetOptOutNodeEncryption()
```
This will unfortunately not work.

`SetOptOutNodeEncryption` is called by the agent during boot up, if it detects that the local node needs to opt-out of encryption. This means that within the cilium-agent process, code can know if the local node has opted out or not, but outside of the cilium process, this function call here will forever return `false` (since there is no code in the process to set it to true).

This means that other processes, such as the cilium client cli, or the cilium-health cli, will not observe the value used by the agent if they read this value here.

What you will have to do instead is read out this value in the agent, and the expose it in the `WireguardStatus model of the Cilium API as described in https://github.com/cilium/cilium/issues/24239

gandro replied:
In addition, the node encryption status is based on both the state of the opt-out flag, and of the state of the `option.Config.EncryptNode` value. The overall logic should be the following (pseudo-code):

```
nodeEncryptionStatus = "Disabled"
if option.Config.EncryptNode {
  if node.GetOptOutNodeEncryption() {
    nodeEncryptionStatus = "OptedOut"
  } else {
    nodeEncryptionStatus = "Enabled"
  }
}
```

romanspb80 replied:
Thank you, Sebastian, for the code review. I've fixed it.

gandro reviewed (changes requested):
Thanks for the PR. Unfortunately, this approach will not work. I recommend implementing the logic in the WireGuard status getter here:

https://github.com/cilium/cilium/blob/02e61a11e0bb9ca7fbf394147606945150c6b816/pkg/wireguard/agent/agent.go#L567

gandro commented on api/v1/models/wireguard_interface.go:
```diff
@@ -38,6 +38,9 @@ type WireguardInterface struct {
 	PublicKey string `json:"public-key,omitempty"`
+
+	// Node Encryption status
+	NodeEncryption string `json:"node-encryption,omitempty"`
```
It seems like this was added without changing the swapper API. Please apply this modification to 

https://github.com/cilium/cilium/blob/02ef3417903e5acbd0d8cddd7ef3248bc366569e/api/v1/openapi.yaml#L3380-L3385 

And then re-generate this file via `make generate-api`.

gandro reviewed (commented):
Thanks! This is looking much better. There are two issues with the API definition that needs to be addressed, but the high-level approach looks good to me.

gandro commented on pkg/client/client.go:
```diff
@@ -689,8 +689,8 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetai
 			for _, i := range wg.Interfaces {
-				iface := fmt.Sprintf("%s (Pubkey: %s, Port: %d, Peers: %d)",
-					i.Name, i.PublicKey, i.ListenPort, i.PeerCount)
+				iface := fmt.Sprintf("NodeEncryption: %s, %s (Pubkey: %s, Port: %d, Peers: %d)",
```
This adds the field per device, which I don't think is useful. Please add the NodeEncryption status to the top-level `models.WireguardStatus` object. It's a global property and so it doesn't make sense to repeat it for each interface.

romanspb80 replied:
I fixed it. Thanks

gandro commented on pkg/client/client.go:
```diff
@@ -686,16 +686,15 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetai
+				fmt.Fprintf(tab, "  %s\t  %s\t%d\t%d\n", i.Name, i.PublicKey, i.ListenPort, i.PeerCount)
 			}
-			fields = append(fields, fmt.Sprintf("[%s]", strings.Join(ifaces, ", ")))
+			tab.Flush()
```
It seems those changes here changed the output. It now uses two lines instead of one, which I don't think is what we want for non-verbose mode:

Current output (cilium status, without --verbose):
```
Encryption:              Wireguard,   (NodeEncryption: Enabled)
  Interface      Pubkey                                         Port    Peers
  cilium_wg0     39nT9TUaYVGFFw5yBZI3qs/iMEgOQ6lqUSrVSVblsEk=   51871   2
```

Expected output:
```
Encryption:              Wireguard [NodeEncryption: Enabled, cilium_wg0 (Pubkey: 39nT9TUaYVGFFw5yBZI3qs/iMEgOQ6lqUSrVSVblsEk=, Port: 51871, Peers: 2)]
```

I would only expect the current multi-line output in `--verbose` mode.

romanspb80 replied:
Ok. I returned the one-line output.

gandro replied:
Thanks!

gandro reviewed (changes requested):
Thanks! Looks good in terms of API, but local testing revealed that the formatting seem off now.

gandro commented on pkg/client/client.go:
```diff
@@ -687,15 +687,16 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetai
 		}
 
-		fmt.Fprintf(w, "Encryption:\t%s\n", strings.Join(fields, "\t"))
+		fmt.Fprintf(w, "Encryption:\t%s\n", strings.Join(fields, ", "))
```
I think we should keep the tab between the state (i.e. Encryption.Mode) and the details. We're not super consistent in the output, but other subsystems (e.g. Hubble, Kubernetes) also seem to use a tab instead of a comma.

Therefore, I'd propose something like this:
```go
fmt.Fprintf(w, "Encryption:\t%s\t%s\n", sr.Encryption.Mode, strings.Join(fields, ", "))
```

(and of course remove `sr.Encryption.Mode` from the `fields` array on line 685)

romanspb80 replied:
> I think we should keep the tab between the state (i.e. Encryption.Mode) and the details. We're not super consistent in the output, but other subsystems (e.g. Hubble, Kubernetes) also seem to use a tab instead of a comma.
> 
> Therefore, I'd propose something like this:
> 
> ```go
> fmt.Fprintf(w, "Encryption:\t%s\t%s\n", sr.Encryption.Mode, strings.Join(fields, ", "))
> ```
> 
> (and of course remove `sr.Encryption.Mode` from the `fields` array on line 685)

Ok. I fixed it
Thanks

gandro reviewed (commented):
Awesome, thanks! One last nit, then this should be good to go.

romanspb80 committed:
Add encrypt-node status to cilium status

This patch adds param "NodeEncryption" to the cilium status

$ cilium status
[...]
Encryption:              Wireguard   [NodeEncryption: Enabled, cilium_wg0 (Pubkey: IoCvq0AB81XznieMpRyBl4XRcRxFuaN3Xhqefun5YQA=, Port: 51871, Peers: 0)]
[...]

Fixes: #24239

Signed-off-by: Roman Ptitcyn <romanspb@yahoo.com>

gandro reviewed (approved):
Awesome, thanks a lot for tackling this!
